### PR TITLE
[Uid] Optimize UuidV4 generation

### PR DESCRIPTION
**Pull request by smnandre**
| Q             | A
| ------------- | ---
| Branch?       | 7.1
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Issues        | Fix #
| License       | MIT

I tried to optimize as much as I could the [UuidV4 generation](https://github.com/symfony/symfony/blob/7.1/src/Symfony/Component/Uid/UuidV4.php).

According to local tests, the code is now between `35%` and `40%` faster.

(But to be fair... we are talking nanoseconds here 😅)

Thanks @nicolas-grekas for the help in this ~~useless~~ fun quest  😃 

-- 

--> Bench (& results): https://github.com/smnandre/uuid-bench

### CODE REVIEW

**Comment by lyrixx** [comment]
I would be nice to add more comments in the code to explain the code

**Comment by smnandre** [comment]
@lyrixx 
I added comments to explain step by step what's done. Tell me if you see anything incorrect there :)

File: src/Symfony/Component/Uid/UuidV4.php [hunk 1]
```diff
@@ -23,12 +23,19 @@ class UuidV4 extends Uuid
     public function __construct(?string $uuid = null)
     {
         if (null === $uuid) {
-            $uuid = random_bytes(16);
-            $uuid[6] = $uuid[6] & "\x0F" | "\x40";
-            $uuid[8] = $uuid[8] & "\x3F" | "\x80";
-            $uuid = bin2hex($uuid);
-
-            $this->uid = substr($uuid, 0, 8).'-'.substr($uuid, 8, 4).'-'.substr($uuid, 12, 4).'-'.substr($uuid, 16, 4).'-'.substr($uuid, 20, 12);
+            // Generate 36 random hex characters (128 bits)
```
Comment (stof) [inline src/Symfony/Component/Uid/UuidV4.php]: This comment is a lie. You are generating more than 128 bits of randomness. 18 bytes is 144 bits. Generating 128 bits was what the old code was doing.

Comment (smnandre) [reply src/Symfony/Component/Uid/UuidV4.php]: A _lie_  😱  //   It's fixed thank you :)

**Review by nicolas-grekas — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @smnandre.

**Comment by rafageist** [comment]
Hi, @fabpot, I've made some additional improvements in the new PR #54027. Please take a look and provide your feedback. Thanks!

**Pull request merged**